openedx/core/djangoapps/content/search/management/commands/reindex_studio.py

```diff
@@ -0,0 +1,140 @@
+        index_name = settings.MEILISEARCH_INDEX_PREFIX + STUDIO_INDEX_NAME
+        with self.using_temp_index(index_name) as temp_index_name:
+            ############## Configure the index ##############
+            # Mark usage_key as unique (it's not the primary key for the index, but nevertheless must be unique):
+            client.index(temp_index_name).update_distinct_attribute(Fields.usage_key)
+            # Mark which attributes can be used for filtering/faceted search:
+            client.index(temp_index_name).update_filterable_attributes([
+                Fields.block_type,
+                Fields.context_key,
+                Fields.org,
+                Fields.tags,
+                Fields.type,
+            ])
+            # Mark which attributes are used for keyword search, in order of importance:
+            client.index(temp_index_name).update_searchable_attributes([
+                Fields.display_name,
+                Fields.block_id,
+                Fields.content,
+                Fields.tags,
+                # Keyword search does _not_ search the course name, course ID, breadcrumbs, block type, or other fields.
+            ])
+
+            ############## Libraries ##############
+            self.stdout.write("Indexing libraries...")
+            for lib_key in lib_keys:
+                self.stdout.write(f"{num_contexts_done + 1}/{num_contexts}. Now indexing library {lib_key}")
+                docs = []
+                for component in lib_api.get_library_components(lib_key):
+                    metadata = lib_api.LibraryXBlockMetadata.from_component(lib_key, component)
+                    doc = searchable_doc_for_library_block(metadata)
+                    docs.append(doc)
+                    num_blocks_done += 1
+                if docs:
+                    # Add all the docs in this library at once (usually faster than adding one at a time):
+                    self.wait_for_meili_task(client.index(temp_index_name).add_documents(docs))
+                self.wait_for_meili_task(client.index(temp_index_name).add_documents(docs))
+                num_contexts_done += 1
+
+            ############## Courses ##############
+            self.stdout.write("Indexing courses...")
+            for course in all_courses:
+                self.stdout.write(
+                    f"{num_contexts_done + 1}/{num_contexts}. Now indexing course {course.display_name} ({course.id})"
+                )
+                docs = []
+
+                def add_with_children(block):
+                    """ Recursively index the given XBlock/component """
+                    doc = searchable_doc_for_course_block(block)
+                    docs.append(doc)  # pylint: disable=cell-var-from-loop
+                    self.recurse_children(block, add_with_children)  # pylint: disable=cell-var-from-loop
+
+                self.recurse_children(course, add_with_children)
```

Comment:
Fetching children at each step like this in split mongo is likely to be painfully slow, and we can't prefetch them with `depth=None` on the get_course call without exploding memory. So I think the best way to do this might be to iterate the CourseOverview ids to get the course keys, and then do a separate `get_course(course_key, depth=None)` to prefetch the children.

Reply:
I added `get_course(..., depth=None)` in 1ff4b756074de333e31bb2b041c17f7c6612d52a . Is that sufficient, or do you want me to change to loading the IDs from CourseOverview too? (I couldn't see a public API for "get all courses" in [the CourseOverview API](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/content/course_overviews/api.py) so I just left it for now.)

This didn't make any noticeable difference on my devstack but presumably that's because I don't have [large] enough courses.

Reply:
I think that's fine for now, thank you.